pkg/internalapis/featurestates/featurestates.go

```diff
@@ -0,0 +1,438 @@
+	}
+}
+
+// configMapDeleted is called when config-map is deleted
+func configMapDeleted(obj interface{}) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	ctx = logger.NewContextWithLogger(ctx)
+	log := logger.GetLogger(ctx)
+
+	fssConfigMap, ok := obj.(*v1.ConfigMap)
+	if fssConfigMap == nil || !ok {
+		log.Warnf("configMapDeleted: unrecognized object %+v", obj)
+		return
+	}
+	if fssConfigMap.Name == supervisorFeatureStatConfigMapName &&
+		fssConfigMap.Namespace == supervisorFeatureStateConfigMapNamespace {
+		log.Warnf("supervisor feature switch state configmap %q from the namespace: %q is deleted", supervisorFeatureStatConfigMapName, supervisorFeatureStateConfigMapNamespace)
+	}
+}
+
+// namespaceAdded adds is called when new namespace is added on the k8s cluster.
+func namespaceAdded(obj interface{}) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	ctx = logger.NewContextWithLogger(ctx)
+	log := logger.GetLogger(ctx)
+
+	namespace, ok := obj.(*v1.Namespace)
+	if namespace == nil || !ok {
+		log.Warnf("namespaceAdded: unrecognized object %+v", obj)
+		return
+	}
+	if _, ok = namespace.Labels[WorkLoadNamespaceLabelKey]; ok {
+		log.Infof("observed new workload namespace: %v", namespace.Name)
+		pendingCRUpdatesObj.enqueueFeatureStateUpdatesForWorkloadNamespace(ctx, namespace.Name)
+	}
+}
+
+// namespaceUpdated is called when namespace is updated
+func namespaceUpdated(oldObj, newObj interface{}) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	ctx = logger.NewContextWithLogger(ctx)
+	log := logger.GetLogger(ctx)
+
+	oldNamespace, ok := oldObj.(*v1.Namespace)
+	if oldNamespace == nil || !ok {
+		log.Warnf("namespaceUpdated: unrecognized object %+v", oldObj)
+		return
+	}
+	newNamespace, ok := newObj.(*v1.Namespace)
+	if newNamespace == nil || !ok {
+		log.Warnf("namespaceUpdated: unrecognized object %+v", oldObj)
+		return
+	}
+
+	_, labelPresentInOldNamespace := oldNamespace.Labels[WorkLoadNamespaceLabelKey]
+	_, labelPresentInNewNamespace := newNamespace.Labels[WorkLoadNamespaceLabelKey]
+
+	if !labelPresentInOldNamespace && labelPresentInNewNamespace {
+		log.Infof("observed new workload namespace: %v", newNamespace.Name)
+		pendingCRUpdatesObj.enqueueFeatureStateUpdatesForWorkloadNamespace(ctx, newNamespace.Name)
+	}
+}
+
+// namespaceDeleted is called when namespace is deleted
+func namespaceDeleted(obj interface{}) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	ctx = logger.NewContextWithLogger(ctx)
+	log := logger.GetLogger(ctx)
+
+	namespace, ok := obj.(*v1.Namespace)
+	if namespace == nil || !ok {
+		log.Warnf("namespaceDeleted: unrecognized object %+v", obj)
+		return
+	}
+	log.Infof("namespace: %q is deleted", namespace.Namespace)
+
+	pendingCRUpdatesObj.lock.Lock()
+	defer pendingCRUpdatesObj.lock.Unlock()
+	delete(pendingCRUpdatesObj.namespaceUpdateMap, namespace.Name)
```

Comment:
Looks like we are deleting the namespace from the map only on deletion of the SV namespace. Does it mean we will keep it in memory for as long as the SV namespace is present? If yes, why?

Reply:
Yes we delete the namespace key from namespaceUpdateMap only when we observe namespace is deleted. The reason is when FSS update is required, we do not require to query, kuberetes to get list of all workload namespaces to push updates.